apollo-configservice/src/main/java/com/ctrip/framework/apollo/configservice/controller/ConfigController.java

```diff
@@ -144,8 +162,52 @@ public ApolloConfig queryConfig(@PathVariable String appId, @PathVariable String
     }
 
     ApolloConfig apolloConfig = new ApolloConfig(appId, appClusterNameLoaded, originalNamespace,
-        mergedReleaseKey);
-    apolloConfig.setConfigurations(mergeReleaseConfigurations(releases));
+        latestMergedReleaseKey);
+
+    Map<String, String> latestConfigurations = mergeReleaseConfigurations(releases);
+
+    try {
+      if (bizConfig.isConfigServiceIncrementalChangeEnabled()) {
+        LinkedHashSet<String> clientSideReleaseKeys = Sets.newLinkedHashSet(
+            Arrays.stream(
+                    clientSideReleaseKey.split(Pattern.quote(ConfigConsts.CLUSTER_NAMESPACE_SEPARATOR)))
+                .collect(Collectors.toList()));
+
+        Map<String, Release> clientSideReleases = configService.findReleasesByReleaseKeys(
+            clientSideReleaseKeys);
+        //find history releases
+        if (clientSideReleases != null) {
+          //order by clientSideReleaseKeys
+          List<Release> historyReleasesWithOrder = new ArrayList<>();
+          for (String item : clientSideReleaseKeys) {
+            Release release = clientSideReleases.get(item);
+            if (release != null) {
+              historyReleasesWithOrder.add(release);
+            }
+          }
+
+          Map<String, String> clientSideConfigurations = mergeReleaseConfigurations
```

Comment:
If `clientSideReleases` is empty (e.g., `-1` on the first client load), it's better to fall back to full sync.